drivers/espi/espi_taf_npcx.c

```diff
@@ -0,0 +1,470 @@
+		taf_head.type = cycle_type;
+		taf_head.tag_hlen = (taf_data_ptr->tag << 4) | ((len & 0xF00) >> 8);
+		taf_head.llen = len & 0xFF;
+		memcpy(data_ptr, &taf_head, sizeof(taf_head));
+
+		rc = flash_read(spi_dev, addr, data_ptr + 4, len);
+		if (rc) {
+			LOG_ERR("flash read fail 0x%x", rc);
+			return -EIO;
+		}
+
+		taf_npcx_completion_handler(dev, (uint32_t *)taf_data_ptr->data);
+
+		total_len -= len;
+		addr += len;
+
+		if (total_len <= config->rx_plsz) {
+			cycle_type = CYC_SCS_CMP_WITH_DATA_LAST;
+			len = total_len;
+		} else {
+			cycle_type = CYC_SCS_CMP_WITH_DATA_MIDDLE;
+		}
+	} while (total_len);
+
+	return 0;
+}
+
+int espi_taf_npcx_flash_write(const struct device *dev, struct espi_saf_packet *pckt)
+{
+	struct espi_taf_npcx_pckt *taf_data_ptr = (struct espi_taf_npcx_pckt *)pckt->buf;
+	uint8_t *data_ptr = (uint8_t *)(taf_data_ptr->data);
+	struct npcx_taf_head taf_head;
+	uint16_t len;
+	int rc;
+
+	if (espi_taf_check_write_protect(dev, pckt->flash_addr,
+					 pckt->len, taf_data_ptr->tag)) {
+		LOG_ERR("Access protection region");
+		return -EINVAL;
+	}
+
+	rc = flash_write(spi_dev, pckt->flash_addr, data_ptr, pckt->len);
+	if (rc) {
+		LOG_ERR("flash write fail 0x%x", rc);
+		return -EIO;
+	}
+
+	/* Completion without Data the length field must be driven to zeros */
+	len = 0;
+
+	taf_head.pkt_len = len + 3;
+	taf_head.type = CYC_SCS_CMP_WITHOUT_DATA;
+	taf_head.tag_hlen = (taf_data_ptr->tag << 4) | ((len & 0xF00) >> 8);
+	taf_head.llen = len & 0xFF;
+	memcpy(data_ptr, &taf_head, sizeof(taf_head));
+
+	taf_npcx_completion_handler(dev, (uint32_t *)taf_data_ptr->data);
+	return 0;
+}
+
+int espi_taf_npcx_flash_erase(const struct device *dev, struct espi_saf_packet *pckt)
+{
+	struct espi_taf_npcx_pckt *taf_data_ptr = (struct espi_taf_npcx_pckt *)pckt->buf;
+	uint8_t *data_ptr = (uint8_t *)taf_data_ptr->data;
+	uint32_t addr = pckt->flash_addr;
+	uint32_t len = pckt->len;
+	struct npcx_taf_head taf_head;
+	int rc;
+
+	if (espi_taf_check_write_protect(dev, addr, len, taf_data_ptr->tag)) {
+		LOG_ERR("Access protection region");
+		return -EINVAL;
+	}
+
+	rc = flash_erase(spi_dev, addr, len);
+	if (rc) {
+		LOG_ERR("flash erase fail");
+		return -EIO;
+	}
+
+	/* Completion without Data the length field must be driven to zeros */
+	len = 0;
+
+	taf_head.pkt_len = len + 3;
+	taf_head.type = CYC_SCS_CMP_WITHOUT_DATA;
+	taf_head.tag_hlen = (taf_data_ptr->tag << 4) | ((len & 0xF00) >> 8);
+	taf_head.llen = len & 0xFF;
+	memcpy(data_ptr, &taf_head, sizeof(taf_head));
+
+	taf_npcx_completion_handler(dev, (uint32_t *)taf_data_ptr->data);
+	return 0;
+}
+
+int espi_taf_npcx_flash_unsuccess(const struct device *dev, struct espi_saf_packet *pckt)
+{
+	struct espi_taf_npcx_pckt *taf_data_ptr
+			= (struct espi_taf_npcx_pckt *)pckt->buf;
+	uint8_t *data_ptr = (uint8_t *)taf_data_ptr->data;
+	struct npcx_taf_head taf_head;
+	uint16_t len;
+
+	/* Completion without Data the length field must be driven to zeros */
+	len = 0;
```

Comment:
why do you even need `len` here then? (same for few others above)

Reply:
Done.  
The length field of write and erase completion packet is zero. This is because these two requests will not return flash data.  
In addition, the unsuccessful completion packet is the same.
Thus, remove `len` and directly fill zero to length field in packet.